vms/proposervm/height_indexed_vm.go

```diff
@@ -118,14 +119,100 @@ func (vm *VM) storeHeightEntry(height uint64, blkID ids.ID) error {
 		if err := vm.State.SetForkHeight(height); err != nil {
 			return fmt.Errorf("failed storing fork height: %w", err)
 		}
+		forkHeight = height
 
 	default:
 		return fmt.Errorf("failed to load fork height: %w", err)
 	}
 
+	if err := vm.State.SetBlockIDAtHeight(height, blkID); err != nil {
+		return err
+	}
+
 	vm.ctx.Log.Debug("indexed block",
 		zap.Stringer("blkID", blkID),
 		zap.Uint64("height", height),
 	)
-	return vm.State.SetBlockIDAtHeight(height, blkID)
+
+	if vm.numHistoricalBlocks == 0 {
+		return nil
+	}
+
+	blocksSinceFork := height - forkHeight
+	// Note: The last accepted block is not considered a historical block. Which
+	// is why <= is used rather than <. This prevents the user from only storing
+	// the last accepted block, which can never be safe due to the non-atomic
+	// commits between the proposervm database and the innerVM's database.
+	if blocksSinceFork <= vm.numHistoricalBlocks {
+		return nil
+	}
+
+	// Note: heightToDelete is >= forkHeight, so it is guaranteed not to
+	// underflow.
+	heightToDelete := height - vm.numHistoricalBlocks - 1
+	blockToDelete, err := vm.State.GetBlockIDAtHeight(heightToDelete)
+	if err == database.ErrNotFound {
+		// Block may have already been deleted. This can happen due to a
```

Comment:
could we be trying to delete genesis (heightToDelete == 0)? Even is such a case it's fine returning nil. I'd just add this possibility to the comment

Reply:
The proposervm never wraps the genesis block. Also at this point we are guaranteed that `heightToDelete >= forkHeight`

Reply:
Right.